src/clj/kixi/hecuba/data/calculate.clj

```diff
@@ -159,6 +159,66 @@
           (measurements/insert-measurements store calculated-sensor page-size calculated)
           (sensors/update-sensor-metadata session calculated-sensor (:start-date range) (:end-date range)))))))
 
+
+;;;;;;;;;;; Tariff calculations ;;;;;;;;;
+
+(defn min-value
+  "Finds the least of a sequence of numbers.
+  Returns a numerical value."
+  [xs]
+  (when (seq xs)
+    (apply min xs)))
+
+(defn min-reading-for-sequence
+  "Gets a sequence of data for a period of time.
+  Retrieve the numerical values. And returns the minimun."
+  [xs]
+  (->> xs
+       (measurements/parse-measurements)
+       (keep :value)
+       (min-value)))
+
+(defmulti find-calc (fn [calculation data] calculation))
+
+(defmethod find-calc :min [_ data]
+  (min-reading-for-sequence data))
+
+(defn calculate-batch
+  [store {:keys [device_id type sensor_id period]} start-date end-date calculation]
+  (db/with-session [session (:hecuba-session store)]
+    (let [month        (time/get-month-partition-key start-date)
+          where        [[= :device_id device_id] [= :sensor_id sensor_id] [= :month month]
+                        [>= :timestamp start-date] [< :timestamp end-date]]
+          measurements         (db/execute session
+                                   (hayt/select :partitioned_measurements (hayt/where where)))]
+      (when (seq measurements)
+        ;; TODO insert this into C* under a synthetic sensor
+        {:value (str (round (find-calc calculation measurements)))
+         :timestamp (tc/to-date end-date)
+         :year (time/get-year-partition-key start-date)
+         :device_id device_id
+         :sensor_id sensor_id})
+      end-date)))
+
+(defn min-reading-for-a-day
+  "Finds the minimum value for a single day.
+  Works in batches of one day. Result of each batch is
+  inserted onto C*."
+  [store {:keys [sensor range]}]
+  (let [{:keys [start-date end-date]} range]
+    (loop [start  start-date]
+      (when-not (t/before? end-date start)
+        (recur (calculate-batch store sensor start (t/plus start (t/days 1)) :min))))))
+
+(defn min-reading-for-4-weeks
+  "Finds the minimum value for a rolling 4 week window.
+  Works in batches of a day, and inserts found value into C*."
+  [store {:keys [sensor range]}]
+  (let [{:keys [start-date end-date]} range]
+    (loop [start  start-date]
+      (when-not (t/before? end-date start)
+        (recur (calculate-batch store sensor (t/minus start (t/weeks 4)) (t/plus start (t/days 1)) :min))))))
```

Comment:
Create a clj-time date-seq from start-date -> end-date and then (doseq ...) over that rather than loop/recur with the when-not for both min-reading-for-a-day and min-reading-for-4-weeks.

min-reading-for-four-weeks will also need to look for data 4 weeks before the start date to calculate the min for the start-date, but will need to recalculate the mins for 4 weeks *after* the end-date in case the rolling minimum is in the end part of the the dirty range.